File: Servers/LeafwixServer/LeafwixServerBLL/Services/Implementation/PlantService.cs
using LeafwixServerBLL.Models;
using LeafwixServerBLL.Services.Interfaces;
using LeafwixServerDAL.Entities.App;
using LeafwixServerDAL.Enums;
using LeafwixServerDAL.Repositories.Interfaces;
using System.Linq;

namespace LeafwixServerBLL.Services.Implementation
{
    public class PlantService : IPlantService
    {
        private readonly IPlantRepository _repository;

        public PlantService(IPlantRepository repository)
        {
            _repository = repository;
        }

        public async Task<List<PlantListResponse>> GetAllPlantsAsync(Guid userId)
        {
            var plants = await _repository.GetAllPlantsAsync(userId);

            return plants.Select(plant =>
            {
                var status = GetStatus(plant.Age);
                var imageUrl = GetPlantImage(plant, status);

                return new PlantListResponse(plant.Id, plant.Name, plant.Age, imageUrl, plant.PlantSpecies.Name, plant.Health);
            }).ToList();
        }

        public async Task<PlantResponse?> GetPlantAsync(Guid userId, Guid plantId)
        {
            var plant = await _repository.GetPlantAsync(userId, plantId);
            if (plant == null) return null;
            var status = GetStatus(plant.Age);
            var imageUrl = GetPlantImage(plant, status);
            return new PlantResponse(plant.Id, plant.Name, plant.Age, imageUrl, plant.Health);
        }

        public async Task AddPlantAsync(AddPlantRequest addPlantRequest)
        {
            var plant = new Plant
            {
                Id = Guid.NewGuid(),
                Name = addPlantRequest.Name,
                Age = addPlantRequest.Age,
                Health = 100,
                LastWatered = DateTime.Now.ToUniversalTime(),
                PlantSpeciesId = addPlantRequest.PlantSpeciesId,
                UserId = addPlantRequest.UserId
            };

            await _repository.AddAsync(plant);
        }

        public async Task UpdatePlantAsync(UpdatePlantRequest updatePlantRequest)
        {
            var plant = await _repository.GetPlantAsync(updatePlantRequest.UserId, updatePlantRequest.PlantSpeciesId);
            if (plant == null) throw new Exception("Plant not found");

            plant.Name = updatePlantRequest.Name;
            plant.Age = updatePlantRequest.Age;
            plant.Health = updatePlantRequest.Health;

            await _repository.UpdateAsync(plant);
        }

        public async Task DeletePlantAsync(Guid userId, Guid plantId)
        {
            var plant = await _repository.GetPlantAsync(userId, plantId);
            if (plant == null) throw new Exception("Plant not found");

            await _repository.DeleteAsync(plant);
        }

        /* Допоміжні методи */

        private string GetPlantImage(Plant plant, Status status)
        {

            string size = "";
            if (status == Status.Small)
            {
                size = "small";
            }
            else if (status == Status.Medium)
            {
                size = "medium";
            }
            else
            {
                size = "big";
            }

            string species = "";
            if (plant.PlantSpecies.Name == "African Violet (Saintpaulia ionantha)")
            {
                species = "african_violet";
            }
            else if (plant.PlantSpecies.Name == "Cast Iron Plant (Aspidistra elatior)")
            {
                species = "aspidistra_elatior";
            }
            else if (plant.PlantSpecies.Name == "Aloe Vera")
            {
                species = "aloe_vera";
            }
            else if (plant.PlantSpecies.Name == "Dracaena")
            {
                species = "dracaena_fragrans";
            }
            else
            {
                species = "default_plant";
            }

            return $"/{size}_{species}.svg";
        }

        // Визначення статусу здоров'я рослини
        private Status GetStatus(double health)
        {
            if (health <= 3)
            {
                return Status.Small;
            }
            else if (health >= 3 && health < 5)
            {
                return Status.Medium;
            }
            else
            {
                return Status.Big;
            }
        }
    }
}

File: Clients/Web/leafwix/src/services/PlantService.js
import { USE_PLATFORM, Platforms } from "../data/constants"
import { externalAxios } from "../lib/Axios"

// axiosPrivate use baseAxios!!
export const PlantService = {
    errorMessages : {
        400: "Validation error",
        409: "Conflict error",
        500: "Internal server error"
    },

    async getAllSpecies(axiosPrivate) {
        if (USE_PLATFORM == Platforms.Web) {
            try {
                const response = await axiosPrivate.get("api/PlantSpecies/getAll")
                return {
                    isError: false,
                    status: response.status,
                    dataOrError: response.data
                };
            } catch (err) {
                console.log(`PlantService (getAllSpecies) -> Error ${err.response.status}: ${this.errorMessages[err.response.status] || "Unknown error"}`);
                return {
                    isError: true,
                    status: err.response.status,
                    dataOrError: err.response.data.title
                };
            }
        }
    },

    async getAllPlantsByUser(axiosPrivate,requestData) {
        if (USE_PLATFORM == Platforms.Web) {
            try {
                const response = await axiosPrivate.get(`api/Plant/getAll/${requestData.userid}`)
                return {
                    isError: false,
                    status: response.status,
                    dataOrError: response.data
                };
            } catch (err) {
                console.log(`PlantService (getAllSpecies) -> Error ${err.response.status}: ${this.errorMessages[err.response.status] || "Unknown error"}`);
                return {
                    isError: true,
                    status: err.response.status,
                    dataOrError: err.response.data.title
                };
            }
        }
    },

    async addPlant(axiosPrivate, requestData) {
        if (USE_PLATFORM == Platforms.Web) {
            try{
                const response = await axiosPrivate.post("api/Plant/add-plant", {
                    name: requestData.name,
                    age: requestData.age,
                    plantSpeciesId: requestData.plantSpeciesId,
                    userId: requestData.userId
                })
                return {
                    isError: false,
                    status: response.status,
                    dataOrError: response.data
                };
            } catch (err) {
                console.log(`PlantService (addPlant) -> Error ${err.response.status}: ${this.errorMessages[err.response.status] || "Unknown error"}`);
                return {
                    isError: true,
                    status: err.response.status,
                    dataOrError: err.response.data.title
                };
            }
        }
    },

    async addHistory(axiosPrivate, requestData) {
        if (USE_PLATFORM == Platforms.Web) {
            try{
                const response = await axiosPrivate.post("api/PlantCareHistory/add", {
                    careType: requestData.careType,
                    plantId: requestData.plantId,
                    userId: requestData.userId,
                })
                return {
                    isError: false,
                    status: response.status,
                    dataOrError: response.data
                };
            } catch (err) {
                console.log(`PlantService (addHistory) -> Error ${err.response.status}: ${this.errorMessages[err.response.status] || "Unknown error"}`);
                return {
                    isError: true,
                    status: err.response.status,
                    dataOrError: err.response.data.title
                };
            }
        }
    },

    async getHistoryByType(axiosPrivate, requestData) {
        if (USE_PLATFORM == Platforms.Web) {
            try{
                const response = await axiosPrivate.post("api/PlantCareHistory/gethistory", {
                    careType: requestData.careType,
                    userId: requestData.userId
                })
                return {
                    isError: false,
                    status: response.status,
                    dataOrError: response.data
                };
            } catch (err) {
                console.log(`PlantService (getHistoryByType) -> Error ${err.response.status}: ${this.errorMessages[err.response.status] || "Unknown error"}`);
                return {
                    isError: true,
                    status: err.response.status,
                    dataOrError: err.response.data.title
                };
            }
        }
    },

    getPlantView(age,type) {
        let size = ""
        if(age <= 3)
        {
            size = "small"
        } else if (age >= 3 && age < 5) {
            size = "medium"
        } else {
            size = "big"
        }
        
        let species = ""
        if(type == "African Violet (Saintpaulia ionantha)")
        {
            species = "african_violet"
        } 
        else if (type == "Cast Iron Plant (Aspidistra elatior)") 
        {
            species = "aspidistra_elatior"
        }
        else if (type == "Aloe Vera") 
        {
            species = "aloe_vera"
        }
        else if (type == "Dracaena") 
        {
            species = "dracaena_fragrans"
        }
        else  {
            species = "default_plant"
        }
        
        return `${size}_${species}.svg`
    }
}
